### ISSUE TITLE
[bench request]: Add NInfer (Qwen3.8-27B NVFP4 and Qwen3.8-Flash-Next, native option-logit readout; adapter and TypeSafe shim included)

### ISSUE THREAD

**Issue by igorls** [fstandhartinger/jevbench#12]
Please consider two systems served by [NInfer](https://github.com/igorls/ninfer), a from-scratch
C++/CUDA inference engine for NVIDIA Blackwell (`sm_120a`), qualified on Windows and buildable on
Linux. Nothing is trained or fitted to the benchmark: the engine exposes the model's own next-token
distribution over the option letters through its `logprob_candidates` extension, so each decision is
one forward pass and no text is generated. Both use public Qwen weights in NInfer's native `.ninfer`
format:

| system | weights | resident VRAM | recommended GPU |
|---|---|---|---|
| `ninfer-qwen3.8-27b` | [neroued/Qwen3.8-27B-nvfp4-NInfer](https://huggingface.co/neroued/Qwen3.8-27B-nvfp4-NInfer) (NVFP4, 20 GiB, from Qwen/Qwen3.8-27B via unsloth/Qwen3.8-27B-NVFP4) | ~24 GB at the profile below | any Blackwell card with 32 GB: RTX 5090, RTX PRO 4500 Blackwell (your RunPod card) |
| `ninfer-qwen3.8-flash-next` | [igorls/Qwen3.8-Flash-Next-mixed-NInfer](https://huggingface.co/igorls/Qwen3.8-Flash-Next-mixed-NInfer) (NVFP4 experts + FP8 projections + INT4 PLE, 105 GiB on disk; 125B total, 6B active, plus a 51B n-gram embedding and a 4B MTP) | ~80 GB at the profile below | RTX PRO 6000 Blackwell 96 GB (RunPod has it) |

**Configuration** (fork commit `3d6abdf8`, branch `research/qwen4-flash-next`; the engine sources are
those of `a77d4ed5`, which is the build measured below, plus one fix to the shim script):

```sh
git clone https://github.com/igorls/ninfer && cd ninfer && git checkout 3d6abdf8
# Linux: CUDA 13.1+, CMake 3.28+, Ninja; text-only build is enough
cmake -S . -B build -G Ninja -DCMAKE_BUILD_TYPE=Release -DNINFER_BUILD_MEDIA=OFF
cmake --build build -j
# or: docker build -t ninfer .   (the Dockerfile in the repo root builds the same two binaries)

hf download neroued/Qwen3.8-27B-nvfp4-NInfer qwen3_8_27b_nvfp4.ninfer --local-dir models
./build/apps/ninfer-serve models/qwen3_8_27b_nvfp4.ninfer --host 0.0.0.0 --port 8010 \
  --model-id qwen3.8-27b --max-context 8192 --kv-capacity 8192 --max-concurrency 1 \
  --kv-dtype fp8 --desktop-reserve-gib 1

hf download igorls/Qwen3.8-Flash-Next-mixed-NInfer qwen3_8_flash_next_mixed.ninfer --local-dir models
./build/apps/ninfer-serve models/qwen3_8_flash_next_mixed.ninfer --host 0.0.0.0 --port 8010 \
  --model-id qwen3.8-flash-next --max-context 32768 --kv-capacity 65536 --max-concurrency 2 \
  --prefill-chunk 8192 --kv-dtype fp8 --gdn-state-dtype bf16 --spec mtp --draft-tokens 4 \
  --desktop-reserve-gib 1
```

Wait for `server status=ready`; Flash-Next materializes 76 GB of weights and warms a host-side
embedding table, so send one warm-up request before timing. The first request after start is
slower on both.

**Two ways to ask it.** `tools/bench/jevbench/ninfer_native.py` is an adapter in your adapter
contract (one Chat Completions request per decision; happy to open it as a pull request). Or run
`tools/bench/jevbench/typesafe_shim.py --ninfer http://127.0.0.1:8010/v1 --ninfer-model qwen3.8-27b
--port 8000`, which serves `POST /v1/systemone` so your `typesafe` adapter runs unchanged with
`--endpoint http://127.0.0.1:8000 --key-env ''`. Both build the identical prompt: the state, the
instructions and the options as `A. label: description` lines, thinking off, one greedy token, and
the distribution read from the option-letter logits and renormalised over the letters. The
requested model name selects the readout: `ninfer-qwen3.8-27b` is the raw distribution,
`ninfer-qwen3.8-27b-t1.5` divides the option logits by 1.5 before the softmax (calibration only,
argmax unchanged). We ask for both 27B rows, see the disclosure.

**Our numbers on the 231 public items** (one request at a time, argmax over the exact label set,
failures count as wrong, none occurred; the score column uses your v1.2 formula on the tiers we can
see, with the judge tier's weight renormalised away). All rows are the Linux build of commit
`a77d4ed5` from the Dockerfile above, run natively under WSL2 Ubuntu 24.04 on an RTX PRO 6000
Blackwell with the exact serve commands above (Flash-Next's Windows build of the same commit gave
the identical answers and probabilities):

| system | easy 48 | standard 72 | hard 111 | hard ECE | fidelity (1 − TVD, 10 items) | p50 / p95 standard | score (see cost) |
|---|---|---|---|---|---|---|---|
| ninfer-qwen3.8-27b | 1.000 | 0.972 | 0.667 | 0.134 | 0.662 | 0.033 s / 0.055 s | 71.0 at $0.09, 66.6 at $0.20 |
| ninfer-qwen3.8-27b-t1.5 | 1.000 | 0.972 | 0.667 | 0.091 | 0.693 | same | 72.4 at $0.09, 68.0 at $0.20 |
| ninfer-qwen3.8-flash-next | 1.000 | 0.986 | 0.793 | 0.096 | 0.826 | 0.094 s / 0.120 s | 71.5 at $0.15 |
| Jev 1.13.0, your per-task file, same items | 1.000 | 0.986 | 0.730 | | | | |

Latencies are localhost wall time on our box; yours will include the network path. Two cost bases
because we do not know which you will apply: your SimpleJev Qwen3.8-27B row used a $0.09 per million
size-class reference, and OpenRouter now lists `qwen/qwen3.8-27b` itself at $0.20 per million input
tokens (there is also a `:free` listing, which we understand you do not price as free).
`qwen/qwen3.8-flash` is $0.15 per million input tokens at its only provider, Alibaba Cloud. Every
decision reads about 685 input tokens and generates nothing.

**Disclosure.**

- Nothing was trained. The prompt wording was fixed before the first run and never changed.
- The temperature of 1.5 for the 27B was chosen on these public hard items after the raw run, by
  sweeping 0.7 to 4 and reading ECE and fidelity, so treat its calibration as a development-set
  number. We declare it here before your run rather than applying it afterwards, and we ask for the
  raw row beside it. For Flash-Next the same sweep found 1.0 already optimal, so it has one row.
- Option order. Our adapter presents choice options in the dataset's `labels` order; TypeSafe's
  wire format carries the criteria object, which your files store in key order, and 119 of the 139
  public choice items differ between the two. Through the shim (wire order) Flash-Next scores hard
  0.775 instead of 0.793, ECE 0.089, fidelity 0.750, with 11 argmax flips, all near 0.5. If you
  run through the shim, expect that variant.
- Numerics. Read-only one-shot requests take a single prefill pass in this commit (publishing
  chat requests keep the split at turn boundaries). The option logits on borderline hard items
  move with execution decomposition and build at the few-percent level: the Windows build of the
  same commit answered 78 of the 111 public hard items where the Linux build answered 74, with six
  argmax flips between them and identical probabilities on most items. Expect the held-out hard
  items to come in below the public numbers, as everyone else's have.
- The public items were consulted repeatedly during development; the held-out ones are yours.

If anything here is unclear or something does not build on your side, ask.

**Comment by igorls** [comment]
One point about the Cost axis for the two rows above, offered as a proposal rather than a complaint.

The rule that prices a self-hosted system at the hosted list tariff of the same weights measures the hosting provider's stack, not the entrant's. Two systems on identical weights get identical cost axes however efficiently they serve, which removes exactly what an inference engine contributes. And for self-hosted rows you do have a real bill: you rent the RunPod card, and you measure the per-decision wall time yourself. GPU seconds times the card's on-demand rate is what the run actually cost, with no invented tariff, and it agrees with your own principle that unmetered is not free.

From our runs, one request at a time and no batching (a conservative basis, since a deployment runs decisions concurrently), the engine is busy for a mean of 85 ms per decision on the 27B and 134 ms on Flash-Next over the 231 public items. At RunPod's posted on-demand rates (read 2026-09-21), and with the other three axes as in the table above:

| system | GPU and rate | $ per 1,000 decisions | cost axis | score (public subset) |
|---|---|---|---|---|
| Jev 1.13.0, its tariff | | $0.0399 | 52.0 | 75.4 (your figure, full set) |
| ninfer-qwen3.8-27b | RTX 5090, $0.69/h community | $0.0163 | 63.6 | 76.9 |
| ninfer-qwen3.8-27b | RTX 5090, $0.99/h secure | $0.0234 | 58.9 | 75.4 |
| ninfer-qwen3.8-27b-t1.5 | RTX 5090, $0.69/h | $0.0163 | 63.6 | 78.4 |
| ninfer-qwen3.8-27b-t1.5 | RTX 5090, $0.99/h | $0.0234 | 58.9 | 76.9 |
| ninfer-qwen3.8-flash-next | RTX PRO 6000, $2.09/h secure | $0.0780 | 43.2 | 73.0 |

Against the tariff basis in the table above ($0.20 per million for the 27B, $0.15 for Flash-Next) that is a swing of 10 to 24 points on the cost axis for the same measurements. Flash-Next stays below Jev on cost either way, because a 96 GB card is expensive and its decisions take longer; the 27B ties or passes Jev.

We are not asking you to replace the tariff basis: it keeps closed APIs and open weights on one scale, and a rented card at full utilisation is not what a customer pays. The ask is to publish a measured-compute basis beside it for self-hosted rows, computed from your own run: mean wall time per decision times the on-demand rate of the card you ran it on. Every input to it is already in your artifact except the hourly rate.

**Comment by fstandhartinger** [comment]
Thanks — this is logged for reachability checks and the JevBench measurement queue. We’ll only make benchmark claims after our own run.

- Harold

**Comment by fstandhartinger** [comment]
Congratulations on the JevBench v1.4 release, Igor L. / NInfer contributors! Thanks for building and sharing your system.

- NInfer Qwen3.8-Flash-Next mixed: #26 at 33.98
- NInfer Qwen3.8-27B NVFP4 (T=1.5): #35 at 26.92
- NInfer Qwen3.8-27B NVFP4: #36 at 26.30

NInfer Qwen3.8-Flash-Next mixed is newly public in v1.4; NInfer Qwen3.8-27B NVFP4 (T=1.5) is newly public in v1.4; NInfer Qwen3.8-27B NVFP4 is newly public in v1.4. V1.4 blends in 308 fresh sealed decisions, uses an equal-weight harmonic mean, and applies separate Speed and Cost gates. For your leading row, sealed accuracy was 34.1% versus 89.6% on the public slice. The >25-point gap adjustment applied; the below-50 Intelligence/Cost gates also applied.

Full board and method: https://benchmarkheaven.com/jev-models
